src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -101,13 +101,45 @@ public class TemplateBinder
             // Any default that doesn't have a corresponding parameter is a 'filter' and if a value
             // is provided for that 'filter' it must match the value in defaults.
             var filters = new RouteValueDictionary(_defaults);
-            foreach (var parameter in _pattern.Parameters)
+            for (var i = 0; i < pattern.Parameters.Count; i++)
             {
-                filters.Remove(parameter.Name);
+                filters.Remove(pattern.Parameters[i].Name);
             }
-
             _filters = filters.ToArray();
-            
+
+            // Now create the constraints and parameter transformers from the pattern
+            var constraints = new List<(string parameterName, IRouteConstraint constraints)>();
```

Comment:
@JamesNK - one other thing that I debated here is passing in the constraints/policies to this class. I really  like to keep workers segmented from services/factories. WOuld you have any objection if I rewrote this such that lines 111-139 were in the caller?